p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
-                    }
-                }
-            }
+        if self.peers.remove(&peer).is_some() {
+            log::warn!("Unregistering unknown peer: {peer}");
         }
     }
 
-    /// Returns a receiver for the chainstate `NewTip` events.
-    async fn subscribe_to_chainstate_events(
-        &mut self,
-    ) -> crate::Result<mpsc::UnboundedReceiver<Id<Block>>> {
-        let (tx, rx) = mpsc::unbounded_channel();
+    /// Announces the header of a new block to peers.
+    async fn handle_new_tip(&mut self, block_id: Id<Block>) -> Result<()> {
+        let header = self
+            .chainstate_handle
+            .call(move |c| c.get_block(block_id))
+            .await??
+            // This should never happen because this block has just been produced by chainstate.
+            .expect("A new tip block unavailable")
+            .header()
+            .clone();
+        self.messaging_handle.make_announcement(Announcement::Block(header))
+    }
 
-        let subscribe_func =
-            Arc::new(
-                move |chainstate_event: chainstate::ChainstateEvent| match chainstate_event {
-                    chainstate::ChainstateEvent::NewTip(block_id, _) => {
-                        if let Err(e) = tx.send(block_id) {
-                            log::error!("PubSubMessageHandler closed: {e:?}")
-                        }
-                    }
-                },
-            );
+    /// Sends a block to the peer.
+    async fn send_block(&mut self, peer: T::PeerId, block: Id<Block>) -> Result<()> {
+        self.peers
+            .get_mut(&peer)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
+            .num_blocks_to_send -= 1;
+
+        let block = self.chainstate_handle.call(move |c| c.get_block(block)).await??.ok_or(
+            P2pError::ProtocolError(ProtocolError::UnknownBlockRequested),
+        )?;
+        self.messaging_handle
+            .send_message(peer, SyncMessage::BlockResponse(BlockResponse::new(block)))
+    }
 
-        self.chainstate_handle
-            .call_mut(|this| this.subscribe_to_events(subscribe_func))
-            .await
-            .map_err(|_| P2pError::SubsystemFailure)?;
+    // TODO: This shouldn't be public.
+    /// Handles a result of message processing.
+    ///
+    /// There are three possible types of errors:
+    /// - Fatal errors will be propagated by this function effectively stopping the sync manager
+    ///   and the whole p2p subsystem.
+    /// - Non-fatal errors aren't propagated, but the peer score will be increased by the
+    ///   "ban score" value of the given error.
+    /// - Ignored errors aren't propagated and don't affect the peer score.
+    pub async fn handle_result(&mut self, peer: T::PeerId, result: Result<()>) -> Result<()> {
+        let error = match result {
+            Ok(()) => return Ok(()),
+            Err(e) => e,
+        };
 
-        Ok(rx)
-    }
+        match error {
```

Comment:
I don't like this complex result processing. This isn't only hard to read but also qualifies to "playing with fire"...

The whole point of the `BanScore` trait is to be able to delegate all this to a single function call that gives you a number that can be audited...

Reply:
I don't like the complex logic here either, but I still think that three distinct cases should be handled differently. There are a few errors (such as `SubsystemFailure`) that indicate a completely broken state, so there is no reason to continue processing. Also some errors should never occur in normal situations (such as `NoiseHandshakeError` or `DialError`). I suppose that ideally we can use more specific errors where it is appropriate instead of general `P2pError` everywhere. I can try to do that if you like this idea, but I prefer to do that in a separate pull request.